[Driver] Add -static flag for generating static archives

troughton opened the pull request:
Add a `-static` flag to the driver that, when combined with `-emit-library`, generates a static library using the platform archiver.

In the future, this `-static` flag will also be used on Windows to determine whether a Swift module should have DLLImport linkage applied.

compnerd commented on include/swift/Driver/Action.h:
```diff
@@ -48,6 +48,7 @@ class Action {
     AutolinkExtractJob,
     REPLJob,
     LinkJob,
+    ArchiveJob,
```
I would like to get @jrose-apple's opinion on naming these: `DyamicLinkJob` and `StaticLinkJob`

jrose-apple replied:
Yeah, I think I like @compnerd's names better (sans typo).

jrose-apple replied:
Doing it in a separate commit (but still the same PR) might be nice for reviewing purposes, but not necessary.

compnerd commented on lib/Driver/DarwinToolChains.cpp:
```diff
@@ -492,6 +494,35 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+  ArgStringList Arguments;
+
+  // Configure the toolchain.
+  const char *LibTool = "libtool";
```
You should be able to use `ar` to do the operation.  The only time that `libtool` is really required is for fat libraries.

troughton replied:
I don't really have an opinion here, but from the previous review: https://github.com/apple/swift/pull/25088#discussion_r289062890

compnerd replied:
Commented on that - I think that we can safely get away with it with `llvm-ar` at least

jrose-apple replied:
Commented against it. :-(

compnerd replied:
I think that we can get away with it using the stock Xcode tools as well (do we really care about the original AT&T Unix?).

compnerd commented on lib/Driver/DarwinToolChains.cpp:
```diff
@@ -284,6 +284,8 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
     Arguments.push_back("-dylib");
     break;
+  case LinkKind::StaticLibrary:
+    llvm_unreachable("invalid link kind");
```
Rather than "invalid link kind" can you clarify that cannot perform archiving with the dynamic linker?

compnerd commented on lib/Driver/Driver.cpp:
```diff
@@ -2177,6 +2196,19 @@ static StringRef baseNameForImage(const JobAction *JA, const OutputInfo &OI,
     return llvm::sys::path::stem(BaseInput);
+  
+  if (auto link = dyn_cast<ArchiveJobAction>(JA)) {
+    Buffer = Triple.isOSWindows() ? "" : "lib";
```
Actually, please prepend `lib` on Windows for the static archives.

compnerd commented on lib/Driver/Driver.cpp:
```diff
@@ -2177,6 +2196,19 @@ static StringRef baseNameForImage(const JobAction *JA, const OutputInfo &OI,
+    if (Triple.isOSWindows())
+      Buffer.append(".lib");
+    else
+      Buffer.append(".a");
```
Just use a ternary here.

compnerd commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -130,6 +130,8 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
     Arguments.push_back("-shared");
     break;
+  case LinkKind::StaticLibrary:
+    llvm_unreachable("invalid link kind");
```
Similar to the Darwin case

compnerd commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -320,6 +322,50 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
+  assert(AR &&
+         "neither ar nor llvm-ar was not found in the toolchain directory or system path.");
+
+  Arguments.push_back("crs");
```
If `llvm-ar` is available, Please add an explicit `-format GNU` to the invocation.

compnerd commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -320,6 +322,50 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
+  }
+
+  assert(AR &&
+         "neither ar nor llvm-ar was not found in the toolchain directory or system path.");
```
This assert is poor - Release builds will drop it and you get no indication of why the command failed.  Please use a proper diagnostic here.

jrose-apple replied:
We haven't been emitted diagnostics from *ToolChains.cpp. Maybe that should change, but most of the other jobs just fall back to invoking the bare name `"ar"` and getting the "not found" error from the system.

compnerd commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -320,6 +322,50 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
+    }
+  }
+
+  if (AR == nullptr) {
```
You can combine this with the previous clause with an `else`:

```
  if (const char *ar = llvm::sys::findProgramByName("llvm-ar", {toolchainPath}))
    Arguments.push_back(context.Args.MakeArgString(ar));
  else if (const char *ar = llvm::sys::findProgramByName("ar", {toolchainPath}))
    Arguments.push_back(context.Args.MakeArgString(ar));
```

compnerd commented on lib/Driver/UnixToolChains.cpp:
```diff
@@ -320,6 +322,50 @@ toolchains::GenericUnix::constructInvocation(const LinkJobAction &job,
+  addInputsOfType(Arguments, context.InputActions, file_types::TY_Object);
+
+  InvocationInfo II{AR, Arguments};
+  II.allowsResponseFiles = true;
```
mmm I don't think that the traditional archiver permits response files.  This should be equivalent to whether `llvm-ar` is in use.

compnerd commented on lib/Driver/WindowsToolChains.cpp:
```diff
@@ -61,6 +61,8 @@ toolchains::Windows::constructInvocation(const LinkJobAction &job,
     Arguments.push_back("-shared");
     break;
+  case LinkKind::StaticLibrary:
+    llvm_unreachable("invalid link kind");
```
Similar to Darwin and Linux cases

compnerd commented on lib/Driver/WindowsToolChains.cpp:
```diff
@@ -193,3 +195,28 @@ toolchains::Windows::constructInvocation(const LinkJobAction &job,
+  ArgStringList Arguments;
+
+  // Configure the toolchain.
+  const char *Lib = "lib";
```
Please use `llvm-ar` if available, otherwise fall back to `lld-link` (`lld-link -lib`), otherwise try `lib`.  I don't think that we need to bother with `link /LIB` as a means of invoking the librarian (`lib`).

troughton replied:
That seems a lot of complexity – three tools with different invocations. I assume you want `llvm-ar` and `lld-link` for cross-compilation, so would it be enough to just pick one of those and say it's required for cross-compilation?

compnerd replied:
I don't want to depend on `lld-link` yet.  Until we have swift-lld created and tracking, please do not depend on `lld-link`.  However, `lld-link` and `link` should be compatible enough, and `link -lib` is the same as `lib.exe`.  So if you want to just use link to invoke the librarian, I am fine with that for Windows.  You can use either `lld-link` or `link` then to invoke `lib`.

compnerd commented on test/Driver/static-archive.swift:
```diff
@@ -0,0 +1,36 @@
+// INFERRED_NAME_DARWIN: libtool -static
+// INFERRED_NAME_DARWIN:  -o libARCHIVER.a
+// INFERRED_NAME_LINUX:   libARCHIVER.a
+// INFERRED_NAME_WINDOWS: ARCHIVER.lib
```
Please add a test case for the invalid combination (`-emit-executable -static`) as well as tests for the fallback path using mocked toolchains.

troughton replied:
Could you clarify what you mean by "tests for the fallback path using mocked toolchains", or point to existing similar tests? I assume you mean ensure that we use the correct tools for e.g. `lld-link`, but I can't see where that's actually tested for the dynamic linker path – there's a line referencing `config.link` in `lit.cfg` but I'm not sure how that feeds into the tests.

jrose-apple commented on lib/Driver/DarwinToolChains.cpp:
```diff
@@ -492,6 +494,35 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+      context.Args.MakeArgString(context.Output.getPrimaryOutputFilename()));
+
+  InvocationInfo II{LibTool, Arguments};
+  II.allowsResponseFiles = true;
```
I don't think libtool does allow response files. It *does* have a form of filelist, though, which you could borrow from the dynamic linking logic on Darwin.

jrose-apple commented on lib/Driver/Driver.cpp:
```diff
@@ -1550,7 +1561,8 @@ void Driver::buildOutputInfo(const ToolChain &TC, const DerivedArgList &Args,
     OI.ModuleName = llvm::sys::path::stem(A->getValue());
-    if (OI.LinkAction == LinkKind::DynamicLibrary &&
+    if ((OI.LinkAction == LinkKind::DynamicLibrary ||
+        OI.LinkAction == LinkKind::StaticLibrary) &&
```
Nitpick: indentation should align by the open-paren.

troughton committed:
Driver: renamic LinkJob to DynamicLinkJob and ArchiveJob to StaticLinkJob

# Conflicts:
#	lib/Driver/DarwinToolChains.cpp

troughton committed:
Driver: add -static flag for generating static archives

troughton commented:
I've removed the parts of this PR which used non-system-default linkers. @compnerd, I know that won't meet your needs, but it's probably best to get the uncontroversial parts of this PR in before dealing with the complexity of customising the choice of linker.

Provided this looks okay with you, @jrose-apple, this should be okay to merge. (I've rebased off master but can't build locally since I don't have Catalina installed – hopefully all the tests still pass CI.)

jrose-apple commented:
I'll take a look soon, but note that you shouldn't need Catalina installed to build, just Xcode 11.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@compnerd, are you okay with this base thing going in, and we can discuss libtool/ar/llvm-ar later?

compnerd commented:
@jrose-apple - okay, lets do that.  At least we have a starting point then to make improvements.
